### PR TITLE
LPS-91190 StatsRequest and StatsResponse SME draft

### PR DESCRIPTION
**Pull request by lipusz**
Draft is ready to be used to write initial article.

### CODE REVIEW

**Comment by rbohl** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: draft.markdown [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+To extend Liferay's search aggregation capabilities, in @product-ver@, we added support for Cardinality Aggregations which provides an approximate (aka. statistical) count of distinct values. Please refer to the "External References" above for more details.
+
+As part of the modularization efforts, [StatsRequest](https://github.com/liferay/liferay-portal/blob/7.2.x/modules/apps/portal-search/portal-search-api/src/main/java/com/liferay/portal/search/stats/StatsRequest.java) and [StatsResponse](https://github.com/liferay/liferay-portal/blob/7.2.x/modules/apps/portal-search/portal-search-api/src/main/java/com/liferay/portal/search/stats/StatsResponse.java) were introduced in the `com.liferay.portal.search.api` module to avoid modifying `portal-kernel`. StatsRequest provides the same statistical features that the legacy `com.liferay.portal.kernel.search.Stats` does besides the new "cardinality" option.
```
Comment (rbohl) [inline draft.markdown]: @lipusz, are we saying here that the StatsRequest and StatsResponse replaces GroupBy and Stats?

File: draft.markdown [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+* SearchSearchRequest#getStats()
+* SearchSearchRequest#setStats(Map<String, Stats> stats)
+
+----
+
+## Draft
+
+Support for [GroupBy](https://github.com/liferay/liferay-portal/blob/7.2.x/portal-kernel/src/com/liferay/portal/kernel/search/GroupBy.java) and [Stats](https://github.com/liferay/liferay-portal/blob/7.2.x/portal-kernel/src/com/liferay/portal/kernel/search/Stats.java) aggregations were introduced in 7.0.
+
+To extend Liferay's search aggregation capabilities, in @product-ver@, we added support for Cardinality Aggregations which provides an approximate (aka. statistical) count of distinct values. Please refer to the "External References" above for more details.
```
Comment (rbohl) [inline draft.markdown]: Cardinality Aggregations sound like facet counts. Could they be used similarly?

For example, could a cardinality agggregation be used to compute the count of returned results matching the tags field value?

File: draft.markdown (hunk 1 above)
Comment (rbohl) [reply draft.markdown]: Maybe a better question is, does anything replace GroupBy?

File: draft.markdown [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+## Search Engine Connector Support
+* Elasticsearch 6: Yes
+* Solr 7: Yes
+
+## New/Related APIs
+
+Listed those which are relevant for developers.
+ 
+API (FQCN) | Provided by Artifact |
```
Comment (rbohl) [inline draft.markdown]: @lipusz, did you right-align the firs column on purpose? I'm trying to decide if I like that format or not :thinking:

File: draft.markdown (hunk 1 above)
Comment (lipusz) [reply draft.markdown]: GroupBy is a TermsAggregation: https://www.elastic.co/guide/en/elasticsearch/reference/6.5/search-aggregations-bucket-terms-aggregation.html

https://github.com/liferay/liferay-portal/blob/master/modules/apps/portal-search-elasticsearch6/portal-search-elasticsearch6-impl/src/main/java/com/liferay/portal/search/elasticsearch6/internal/groupby/DefaultGroupByTranslator.java#L64

So the answer is no, we are not replacing GroupBy, it's a different aggregation type.

File: draft.markdown (hunk 2 above)
Comment (lipusz) [reply draft.markdown]: Facets are "bucket aggregations", cardinality is different. The Elasticsearch cardinality stat guide covers the basics. I'd just redirect the users there.

File: draft.markdown (hunk 3 above)
Comment (lipusz) [reply draft.markdown]: No, I don't know why it happened. Maybe the spaces?

File: draft.markdown (hunk 2 above)
Comment (rbohl) [reply draft.markdown]: I was comparing a facet count to a cardinality aggregations. So a search query looks for 

> localhost:9200/liferay-*/_search?q=title_en_US=test

Couldn't you use a cardinality aggregation to count all the results with entryClassPK=33405 to count all the journal articles matching the query?

File: draft.markdown (hunk 3 above)
Comment (rbohl) [reply draft.markdown]: It was this line:

> ---------: | :------------------: |

That creates a right-aligned column (the first column) and a centered column (the second one). I thought maybe you'd done it on purpose.

File: draft.markdown (hunk 2 above)
Comment (lipusz) [reply draft.markdown]: No. That's two different things.

Faceted search is about executing a "main search" (keyword) then, collect terms from specific fields which represent the "facets" (aka. categories, filtering options etc.) so you can narrow down. And each bucket will have a certain number of docs = facet count.

Cardinality is about performing statistical stuff on top of the results.

File: draft.markdown (hunk 3 above)
Comment (lipusz) [reply draft.markdown]: :-)

Comment (lipusz) [reply draft.markdown]: I still don't understand why it's producing things to be right-aligned... :stuck_out_tongue:

Comment (rbohl) [reply draft.markdown]: | Tables   |      Are      |  Cool |
|----------|:-------------:|------:|
| col 1 is |  left-aligned | $1600 |
| col 2 is |    centered   |   $12 |
| col 3 is | right-aligned |    $1 |

Source: https://www.tablesgenerator.com/markdown_tables

Comment (lipusz) [reply draft.markdown]: Thanks!

Comment (lipusz) [reply draft.markdown]: oh, it's the `:` that controls. Smart. :smirk:

File: draft.markdown [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+```
+8. Perform a search using Searcher and SearchRequest to get `com.liferay.portal.search.searcher.SearchResponse`:
+```java
+
+SearchResponse searcher.search(searchRequest);
+```
+
+**Example:** https://github.com/liferay/liferay-portal/blob/7.2.x/modules/apps/portal-search/portal-search-test-util/src/main/java/com/liferay/portal/search/test/util/stats/BaseStatisticsTestCase.java#L128
+
+**Note:** there is still support for the legacy `com.liferay.portal.kernel.search.Stats` object:
```
Comment (rbohl) [inline draft.markdown]: Do we really need to document the old way?

**Comment by rbohl** [comment]
Superceded: https://github.com/lipusz/liferay-docs/pull/29

**Pull request closed without merging**